README.md
`Horus` [![Build Status](https://travis-ci.org/CotaPreco/Horus.svg?branch=develop)](https://travis-ci.org/CotaPreco/Horus)
=====
Horus is a simple and minimalist event-hub for pipelining events from any direction to the client.

![Horus](https://raw.githubusercontent.com/CotaPreco/Horus/develop/assets/Horus.png "Horus")

- Dependency free: just drop and run; that's why is written in pure **Go**;
- It works very well with what you have today;
- Fast as hell, it does nothing rather than serving a WebSocket and deliver messages;
- Small and tiny, atomized.

# How it works
![How it works](https://raw.githubusercontent.com/CotaPreco/Horus/develop/assets/1.png "How it works")

In few words: your front-end will connect to **Horus** through a [WebSocket](http://en.wikipedia.org/wiki/WebSocket), and start waiting for new messages. And then you have what we call *Receiver*, *Receiver* is basically someone listening for incoming messages to send them to the clients (*...your front-end*).

## Demonstration
[![Horus asciicast](https://asciinema.org/a/19437.png)](https://asciinema.org/a/19437?autoplay=1)

## Install
```
curl -sL https://raw.githubusercontent.com/CotaPreco/Horus/develop/install.sh |sh
```

Or via **wget**:
```
wget -qO- https://raw.githubusercontent.com/CotaPreco/Horus/develop/install.sh |sh
```

And if you're familiar with [Docker](http://www.docker.com/), you can be getting started with:

```
docker run -d -p 8080:8000 -p 7500:7600/udp cotapreco/horus:0.1.0
```

## Usage
```
horus [...opts]
```

#### Available `...opts`:
| Option | Usage | Example | Default
| :---- | :---: | :--- | :---
| `-ws-host` | *Optional* | `-ws-host 127.0.0.1` | 0.0.0.0
| `-ws-port` | *Optional* | `-ws-port 8888` | 8000
| `-receiver-udp-host` | *Optional* | `-receiver-udp-host 127.0.0.1` | 0.0.0.0
| `-receiver-udp-port` | *Optional* | `-receiver-udp-port 5000` | 7600
| `-udp-max-packet-size` | *Optional* | `-udp-max-packet-size 65507` | 8192

## Getting started
```
$ horus -ws-port 8080 -receiver-udp-port 7500
```

Then, **Horus** will be listening for WebSocket connections on `0.0.0.0:8080`, and also there's an [UDP](http://en.wikipedia.org/wiki/User_Datagram_Protocol) receiver at `0.0.0.0:7500` which will wait for messages to deliver.

At this point, you're able to **listen** and **send** messages, if you're familiar with *npm*, you can install a simple tool called [wscat](https://www.npmjs.com/package/wscat):
```
$ npm install -g wscat
```

Then: `wscat -c ws://localhost:8080` and you're ready, listening!

But, you also can simple point your browser to `localhost`, open *Chrome Inspector* or whatever you can type some javascript code and execute:
```JS
var ws = new WebSocket('ws://localhost:8080');

ws.onmessage = function(e) {
  console.log(e.data);
}
```

#### How can I send messages to `UDP` receiver?
If you're using **bash**:
```
$ echo -n "Hello, world" >/dev/udp/0.0.0.0/5000
```

Otherwise, you can go with [netcat](http://en.wikipedia.org/wiki/Netcat):
```
$ echo -n "Hello, world" |nc -4u -w1 0.0.0.0 5000
```

And also there's a list of **known clients**:

| Author | URL
| :----: | :---:
| [@CotaPreco](https://github.com/CotaPreco) | #php [HorusPHPClient](https://github.com/CotaPreco/HorusPHPClient)
| [@CotaPreco](https://github.com/CotaPreco) | #javascript [horusjs](https://github.com/julianocomg/horusjs)

### Will it scale to thousands of connections?
Maybe yes, maybe not. Well, there isn't much to change, you just need to consider increasing the number servers running **Horus**, very simple. And also [HAProxy](http://www.haproxy.org/) can help you deal with that.

# License
[MIT](https://github.com/CotaPreco/Horus/blob/develop/LICENSE) &copy; Cota Preço.


CHANGELOG.md
# Changelog
## 0.1.0
- Allows user to define maximum UDP packet size through `--udp-max-packet-size` (default **8192**) [#22](https://github.com/CotaPreco/Horus/issues/22);
- Removed `EXPOSE` from Dockerfile [#16](https://github.com/CotaPreco/Horus/issues/16);
- Fixed bug when message is `nil` before attempt to `.Send` (TaggedConnectionHub) [#14](https://github.com/CotaPreco/Horus/issues/14);
- Display details on startup (such as WebSocket `address:port` and UdpReceiver `address:port`) [#19](https://github.com/CotaPreco/Horus/issues/19).

## 0.1.0
- `UdpReceiver` & `NullByteReceiveStrategy`;
- `Message` *(usually a broadcast)*, `TaggedMessage` & `TagSequencedMessage` *(strictly must contain all tags)*;
- `Tag` now supports a few more special characters: `A-Z a-z 0-9 - . _ ~ ! $ & ' ( ) * + , ; = : @ / ?`.


main.go
package main

import (
	"flag"
	"fmt"
	"net/http"
	"os"
	"strings"

	"github.com/CotaPreco/Horus/command"
	"github.com/CotaPreco/Horus/receiver/udp"
	"github.com/CotaPreco/Horus/util"
	"github.com/CotaPreco/Horus/ws"
	wsc "github.com/CotaPreco/Horus/ws/command"
	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
	CheckOrigin: func(r *http.Request) bool {
		// @link https://godoc.org/github.com/gorilla/websocket#hdr-Origin_Considerations
		return true
	},
}

var (
	VERSION   = "N/A"
	GITCOMMIT = "N/A"
)

var (
	defaultWsHost           = util.EnvOrDefault("WS_HOST", "0.0.0.0")
	defaultWsPort           = util.EnvOrDefault("WS_PORT", "8000")
	defaultUdpReceiverHost  = util.EnvOrDefault("UDP_RECEIVER_HOST", "0.0.0.0")
	defaultUdpReceiverPort  = util.EnvOrDefault("UDP_RECEIVER_PORT", "7600")
	defaultUdpMaxPacketSize = util.EnvOrDefault("UDP_PACKET_SIZE", "8192")
)

var (
	flgVersion    = flag.Bool("v", false, "")
	udpHost       = flag.String("receiver-udp-host", defaultUdpReceiverHost, "")
	udpPort       = flag.Int("receiver-udp-port", util.Str2int(defaultUdpReceiverPort), "")
	wsHost        = flag.String("ws-host", defaultWsHost, "")
	wsPort        = flag.Int("ws-port", util.Str2int(defaultWsPort), "")
	udpPacketSize = flag.Int("udp-max-packet-size", util.Str2int(defaultUdpMaxPacketSize), "")
)

func main() {
	flag.Usage = func() {
		flag.CommandLine.SetOutput(os.Stdout)

		var help = strings.Trim(`
Horus — A simple and minimalist event-hub for pipelining events :-)

USAGE:
	horus [...OPTIONS]

OPTIONS:
%s
`, "\n")

		var opts string

		for _, opt := range [][]string{
			{
				"-v",
				"Prints the current version of `Horus`",
			}, {
				"-ws-host",
				"Defines in which IP WebSocket will bind to",
			}, {
				"-ws-port",
				"Defines the port for the WebSocket server listen for connections",
			}, {
				"-receiver-udp-host",
				"Defines in which IP the UDP receiver will bind to",
			}, {
				"-receiver-udp-port",
				"Defines the port for receiver listen on",
			}, {
				"-udp-max-packet-size",
				"Defines the maximum buffer size for packet",
			},
		} {
			opts += fmt.Sprintf("\t%-25.20s /* %s */\n", opt[0], opt[1])
		}

		fmt.Printf(help, opts)

		os.Exit(0)
	}

	flag.Parse()

	if *flgVersion {
		fmt.Printf("Horus v%s, build %s\n", VERSION, GITCOMMIT)
		return
	}

	// --
	bus := command.NewGenericCommandBus()
	hub := ws.NewTaggedConnectionHub()

	bus.PushHandler(hub)
	bus.PushHandler(wsc.NewARTagCommandRedispatcher(bus))

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		conn, err := upgrader.Upgrade(w, r, nil)

		if err != nil {
			return
		}

		defer conn.Close()

		hub.Subscribe(conn)

		defer hub.Unsubscribe(conn)

		for {
			messageType, message, err := conn.ReadMessage()

			if err != nil {
				return
			}

			if messageType == websocket.TextMessage {
				bus.Dispatch(wsc.NewSimpleTextCommand(string(message), conn))
			}
		}
	})

	// ---
	receiver := udp.NewUdpReceiver(*udpHost, *udpPort, *udpPacketSize, new(udp.NullByteReceiveStrategy))
	receiver.Attach(hub)

	go receiver.Receive()
	// ---

	fmt.Printf(
		"Udp Receiver — %s:%d\nWebSocket — %s:%d\n",
		*udpHost,
		*udpPort,
		*wsHost,
		*wsPort,
	)

	err := http.ListenAndServe(
		fmt.Sprintf("%s:%d", *wsHost, *wsPort),
		nil,
	)

	util.Invariant(
		err == nil,
		"Unexpected `%s` (ListenAndServe)",
		err,
	)
}
